libs/crypto/merkle/src/main/kotlin/net/corda/crypto/merkle/MerkleProofImpl.kt

```diff
@@ -0,0 +1,115 @@
+     * Then it compares it with the input parameter to check if they match.
+     *
+     * It walks through the levels of the tree from bottom to up following the same logic as the
+     * proof or tree creation. [MerkleTreeImpl.createAuditProof]
+     * It recreates the routes towards the root element from the items in the leaves to be proven with using
+     * the proof's hashes when they are needed.
+     */
+    override fun verify(root: SecureHash, digestProvider: MerkleTreeHashDigestProvider): Boolean {
+        if (leaves.isEmpty()) {
+            return false
+        }
+        if (leaves.any { it.index < 0 || it.index >= treeSize }) {
+            return false
+        }
+        if (leaves.map { it.index }.toSet().size != leaves.size) {
+            return false
+        }
+        var hashIndex = 0
+        val sortedLeaves = leaves.sortedBy { it.index }
+        var nodeHashes = sortedLeaves.map { Pair(it.index, digestProvider.leafHash(it.index, it.nonce, it.leafData)) }
+        var treeDepth = MerkleTreeImpl.treeDepth(treeSize)
+        var currentSize = treeSize
+        while (currentSize > 1) {
+            if (nodeHashes.isEmpty()) {
+                return false
+            }
+            --treeDepth
+            val newItems = mutableListOf<Pair<Int, SecureHash>>()
+            var index = 0
+            while (index < nodeHashes.size) {
+                val item = nodeHashes[index]
+                if (item.first < currentSize and 0x7FFFFFFE) {      // If the level has odd elements, we'll process
+                                                                    // the last element later.
+                    if (index < nodeHashes.size - 1) {              // If there is a next element...
+                        val next = nodeHashes[index + 1]
+                        if (item.first xor next.first == 1) {       // ... and they are a pair with the current
+                            newItems += Pair(                       // in the original tree, we create their parent.
+                                item.first / 2,
+                                digestProvider.nodeHash(treeDepth, item.second, next.second)
+                            )
+                            index += 2
+                            continue
+                        }
+                    }
+                    if (hashIndex >= hashes.size) {                 // We'll need and more hash to continue. So if
+                        return false                                // we do not have more, the proof is incorrect.
+                    }
+                                                                    // We pair the current element with a
+                                                                    // hash from the proof
+                    newItems += if (item.first and 1 == 0) {        // Even index means, that the item is on the left
```

Comment:
```suggestion
                    newItems += if ((item.first and 1) == 0) {        // Even index means, that the item is on the left
```
Would suggest parenthesis to make reading easier - `1 == 0` threw me for a bit.

Reply:
Thanks. Fixed.